### PR TITLE
Upgrade to Java8 and cleaner code

### PR DESCRIPTION
**Pull request by ruiqiding**
- Upgrade to JDK 8.
- Use autovalue to generate valueType classes, which comes with correct hashcode, equals, toString for free. The source code is shorter than JavaBean with getter/setters. It works well with Jackson as well. (AutoValue+Jackson is widely used to build Restful Java services/clients. https://www.baeldung.com/introduction-to-autovalue)
- Removed unnecessary hacks around deserialization
- Fixed bugs in handling connection error and parsing response message.
- Organized classes in Java packages. Rename methods.
- Use correct Exception types
- Replace magic numbers with constants as defined in Http libraeis.
- Update Jackson versions
- Reduced warning/errors reported by IntelliJ
- Removed dead code

### CODE REVIEW

File: src/main/java/com/microsoft/azure/search/samples/client/SearchIndexClient.java [hunk 1]
```diff
@@ -0,0 +1,279 @@
+package com.microsoft.azure.search.samples.client;
```
Comment (bitsforthought) [inline src/main/java/com/microsoft/azure/search/samples/client/SearchIndexClient.java:1]: @ruiqiding It looks like you moved files and changed their content at the same time, which makes review very difficult. Please do these in separate commits so it's easier to review.

**Review by bitsforthought — CHANGES_REQUESTED** [review]

File: src/main/java/com/microsoft/azure/search/samples/client/SearchIndexClient.java [hunk 2]
```diff
@@ -1,137 +1,109 @@
...
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.URLEncoder;
 import java.util.ArrayList;
-import java.util.Collection;
+import java.util.List;
 import java.util.Map;
 
 public class SearchIndexClient {
     private static final String API_VERSION = "2016-09-01";
```
Comment (bitsforthought) [inline src/main/java/com/microsoft/azure/search/samples/client/SearchIndexClient.java]: We could use 2017-11-11 here; It should be backwards compatible.

File: src/main/java/com/microsoft/azure/search/samples/client/SearchIndexClient.java [hunk 3]
```diff
@@ -1,137 +1,109 @@
...
-    }
-
     public SearchIndexClient(String serviceName, String indexName, String apiKey) {
         this.serviceName = serviceName;
         this.indexName = indexName;
         this.apiKey = apiKey;
     }
 
-    public boolean exists() throws IOException {
+    public boolean isIndexExists() throws IOException {
```
Comment (bitsforthought) [inline src/main/java/com/microsoft/azure/search/samples/client/SearchIndexClient.java]: nit: "is" isn't the right verb here. I'd recommend either `doesIndexExist` or `indexExists`.

File: src/main/java/com/microsoft/azure/search/samples/client/SearchIndexClient.java [hunk 4]
```diff
@@ -1,137 +1,109 @@
...
-
-    public boolean delete() throws IOException {
+    public void deleteIndexIfExists() throws IOException {
         HttpURLConnection connection = httpRequest(buildIndexDefinitionUrl(), "DELETE");
-        int response = connection.getResponseCode();
-        if (response == 404) {
-            return false;
-        }
         throwOnHttpError(connection);
-        return true;
```
Comment (bitsforthought) [inline src/main/java/com/microsoft/azure/search/samples/client/SearchIndexClient.java]: It can be useful to know whether the index was actually deleted or didn't exist; Why remove this functionality?

Also, the `IfExists` suffix suggests that the operation will fail if the index doesn't exist, which is not true.

File: src/main/java/com/microsoft/azure/search/samples/client/SearchIndexClient.java [hunk 5]
```diff
@@ -157,165 +128,152 @@ private String buildIndexListUrl() {
...
     }
 
-    private static <T> T withHttpRetry(RetriableHttpOperation<T> r) throws HttpRetryException, IOException {
-        final int RETRIES = 3;
-        int delay = 30000; // 30 secs to start with
-        HttpRetryException exception = null;
-        T result = null;
-        for (int i = 1; ; i++) {
+    private static <T> T withHttpRetry(RetriableHttpOperation<T> r) throws IOException {
+        int maxRetries = 3;
```
Comment (bitsforthought) [inline src/main/java/com/microsoft/azure/search/samples/client/SearchIndexClient.java]: Why not use `final` for constants?

File: src/main/java/com/microsoft/azure/search/samples/demo/DemoOperations.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
+package com.microsoft.azure.search.samples.demo;
```
Comment (bitsforthought) [inline src/main/java/com/microsoft/azure/search/samples/demo/DemoOperations.java:1]: Try to separate code movement from refactoring. They should be separate commits if possible.

**Review by bitsforthought — CHANGES_REQUESTED** [review]

File: src/main/java/com/microsoft/azure/search/samples/client/SearchIndexClient.java (hunk 2 above)
Comment (ruiqiding) [reply src/main/java/com/microsoft/azure/search/samples/client/SearchIndexClient.java]: Was thinking of doing it when adding features to this sample, for example, autocomplete.
I don't mind to do it now.

File: src/main/java/com/microsoft/azure/search/samples/client/SearchIndexClient.java (hunk 3 above)
Comment (ruiqiding) [reply src/main/java/com/microsoft/azure/search/samples/client/SearchIndexClient.java]: Agree. Thanks.

**Review by bitsforthought — APPROVED** [review]
